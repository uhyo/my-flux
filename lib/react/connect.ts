import {
    Component,
    ComponentClass,
    StatelessComponent,
    createElement,
} from 'react';
import {
    Store,
} from '../store';

export interface Selector<State, OriginalProps, ResultProps>{
    (state: State, props: OriginalProps): ResultProps;
}
export interface Dispatcher<Action>{
    <A extends Action>(action: A): Action;
}

export interface ComponentConnectFunction<OriginalProps, AddedProps extends object>{
    (component: ComponentClass<OriginalProps & AddedProps> | StatelessComponent<OriginalProps & AddedProps>): ComponentClass<OriginalProps>;
}
export function connect<State, Action, OriginalProps extends object, AddedByStateProps extends object, AddedByDispatchProps extends object = {}>(
    store: Store<State, Action>,
    mapStateToProps: Selector<State, OriginalProps, AddedByStateProps>,
    mapDispatchToProps?: Selector<Dispatcher<Action>, OriginalProps, AddedByDispatchProps>,
): ComponentConnectFunction<OriginalProps, AddedByStateProps & AddedByDispatchProps>{
    type AddedProps = AddedByStateProps & AddedByDispatchProps;
    return (component: ComponentClass<OriginalProps & AddedProps> | StatelessComponent<OriginalProps & AddedProps>)=>{
        return class extends Component<OriginalProps, AddedProps>{
            protected unsubscribe: undefined | (()=>void);
            constructor(props: OriginalProps){
                super(props);

                // 現在のstateを取得
                const bystate = mapStateToProps(store.getState(), props);
                const bydispatch = (mapDispatchToProps ? mapDispatchToProps(store.dispatch, props) : {}) as AddedByDispatchProps;
                // FIXME
                const added = {
                    ... (bystate as any),
                    ... (bydispatch as any),
                } as Readonly<AddedProps>;

                this.state = added;
                this.unsubscribe = store.subscribe(state=>{
                    const newState = mapStateToProps(state, props);
                    this.setState(newState);
                });
            }
            componentWillUnmount(){
                /* istanbul ignore else */
                if (this.unsubscribe != null){
                    this.unsubscribe();
                    this.unsubscribe = void 0;
                }
            }
            render(){
                // FIXME
                const {
                    children,
                    ... props2,
                } = (this.props as any);
                const props = {
                    ... props2,
                    ... (this.state as any),
                } as OriginalProps & AddedProps;
                // FIXME
                return createElement(component as any, props, children);
            }
        };
    };
}
