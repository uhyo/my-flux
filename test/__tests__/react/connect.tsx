import * as React from 'react';
import {
    create,
} from 'react-test-renderer';
import {
    Reducer,
} from '../../../lib/reducer';
import {
    Store,
} from '../../../lib/store';
import {
    connect,
} from '../../../lib/react/connect';
describe('connect', ()=>{
    interface Action{
        foo: string;
        bar: number;
    }
    interface State{
        foo: string;
        bar: number;
    }
    interface ComponentProps{
        hey: string;
    }

    class C extends React.Component<ComponentProps & State, {}>{
        render(){
            const {
                foo,
                bar,
                hey,
            } = this.props;
            return <div>
                <p>foo: {foo}</p>
                <p>bar: {bar}</p>
                <p>hey: {hey}</p>
            </div>;
        }
    }
    const reducer: Reducer<State, Action> = (_, action)=>{
        return {
            foo: action.foo,
            bar: action.bar,
        };
    };
    const identity = (state: State)=>state;

    it('basic', ()=>{
        const store = new Store(reducer);
        store.dispatch({
            foo: 'FOO',
            bar: 0,
        });

        const Container = connect(
            store,
            identity,
        )(C);

        const component = create(<Container hey="Hey!" />);

        const rendered = component.toJSON();
        expect(rendered).toMatchSnapshot();
    });
    it('reacts to state change', ()=>{
        const store = new Store(reducer);
        store.dispatch({
            foo: 'FOO',
            bar: 0,
        });

        const Container = connect(
            store,
            identity,
        )(C);
        const component = create(<Container hey="Hey!" />);

        expect(component.toJSON()).toMatchSnapshot();

        store.dispatch({
            foo: "風呂",
            bar: 123,
        });

        expect(component.toJSON()).toMatchSnapshot();
    });
    it('calls selector when store changed', ()=>{
        const store = new Store(reducer, {
            foo: 'FOO',
            bar: 0,
        });

        const selector: typeof identity = jest.fn(identity);
        const Container = connect(
            store,
            selector,
        )(C);

        expect(selector).not.toHaveBeenCalled();

        create(<Container hey="Hoy!" />);

        expect((selector as any).mock.calls).toEqual([
            [{
                foo: 'FOO',
                bar: 0,
            }],
        ]);

        store.dispatch({
            foo: 'BOO',
            bar: -5,
        });

        expect((selector as any).mock.calls).toEqual([
            [{
                foo: 'FOO',
                bar: 0,
            }],
            [{
                foo: 'BOO',
                bar: -5,
            }],
        ]);
    });
    it('unmounting', ()=>{
        const store = new Store(reducer, {
            foo: 'FOO',
            bar: 0,
        });

        const selector: typeof identity = jest.fn(identity);
        const Container = connect(
            store,
            selector,
        )(C);

        expect(selector).not.toHaveBeenCalled();

        const component = create(<Container hey="Hoy!" />);

        component.unmount();

        store.dispatch({
            foo: 'BOO',
            bar: -5,
        });

        expect((selector as any).mock.calls).toEqual([
            [{
                foo: 'FOO',
                bar: 0,
            }],
        ]);
    });
});
