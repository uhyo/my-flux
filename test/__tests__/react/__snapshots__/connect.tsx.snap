// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect basic 1`] = `
<div>
  <p>
    foo: 
    FOO
  </p>
  <p>
    bar: 
    0
  </p>
  <p>
    hey: 
    Hey!
  </p>
</div>
`;

exports[`connect reacts to state change 1`] = `
<div>
  <p>
    foo: 
    FOO
  </p>
  <p>
    bar: 
    0
  </p>
  <p>
    hey: 
    Hey!
  </p>
</div>
`;

exports[`connect reacts to state change 2`] = `
<div>
  <p>
    foo: 
    風呂
  </p>
  <p>
    bar: 
    123
  </p>
  <p>
    hey: 
    Hey!
  </p>
</div>
`;
